src/engine/SCons/Script/Main.py

```diff
@@ -703,67 +703,84 @@ def _load_site_scons_dir(topdir, site_dir_name=None):
-                # module's dictionary, so anything that code defines ends
-                # up adding to that module.  This is really short, but all
-                # the error checking makes it longer.
-                try:
-                    m = sys.modules['SCons.Script']
-                except Exception as e:
-                    fmt = 'cannot import site_init.py: missing SCons.Script module %s'
-                    raise SCons.Errors.InternalError(fmt % repr(e))
-                try:
-                    sfx = description[0]
-                    modname = os.path.basename(pathname)[:-len(sfx)]
-                    site_m = {"__file__": pathname, "__name__": modname, "__doc__": None}
-                    re_special = re.compile("__[^_]+__")
-                    for k in list(m.__dict__.keys()):
-                        if not re_special.match(k):
-                            site_m[k] = m.__dict__[k]
-
-                    # This is the magic.
-                    exec(compile(fp.read(), fp.name, 'exec'), site_m)
-                except KeyboardInterrupt:
-                    raise
-                except Exception as e:
-                    fmt = '*** Error loading site_init file %s:\n'
-                    sys.stderr.write(fmt % repr(site_init_file))
-                    raise
-                else:
-                    for k in site_m:
-                        if not re_special.match(k):
-                            m.__dict__[k] = site_m[k]
-            except KeyboardInterrupt:
-                raise
-            except ImportError as e:
-                fmt = '*** cannot import site init file %s:\n'
-                sys.stderr.write(fmt % repr(site_init_file))
-                raise
-        finally:
-            if fp:
-                fp.close()
     if os.path.exists(site_tools_dir):
-        # prepend to DefaultToolpath
         SCons.Tool.DefaultToolpath.insert(0, os.path.abspath(site_tools_dir))
+    if not os.path.exists(site_init_file):
+        return
+
+    if PY3:
+        import importlib.util
+    else:
+        import imp
+    import re
+
+    try:
+        try:
+            m = sys.modules['SCons.Script']
+        except Exception as e:
+            fmt = 'cannot import site_init.py: missing SCons.Script module %s'
+            raise SCons.Errors.InternalError(fmt % repr(e))
+
+        # Load the site_init file into the SCons.Script namespace.
+        # This is a variant on the basic Python import flow except
+        # the globals dict is prepopulated from the SCons.Script
+        # module object; on success Script's globals are refilled
+        # from the site_init globals so it all appears in SCons.Script.
+        if PY3:
+            # find_spec cannot restrict search location, so do manually
+            # we don't want a "hit" from a mod elsewhere in sys.path
+            save_syspath = sys.path
+            sys.path = [sys.path[0]]
+            spec = importlib.util.find_spec(site_init_modname)
```

Comment:
I can do this even more simply by using `importlib.spec_from_file_location` (don't have to fiddle the path while doing it) if that's preferred.  Tests come out the same.